src/Microsoft.OData.Core/ODataBatchReaderStream.cs

```diff
@@ -343,308 +79,45 @@ internal int ReadWithDelimiter(byte[] userBuffer, int userBufferOffset, int coun
-                        Debug.Assert(lineEndEndPosition < this.batchBuffer.CurrentReadPosition + this.batchBuffer.NumberOfBytesInBuffer, "Line end must finish withing buffer range.");
-
-                        byteCount = lineEndStartPosition - this.batchBuffer.CurrentReadPosition;
-                        if (byteCount > 0)
-                        {
-                            ODataBatchUtils.EnsureArraySize(ref bytesForString, lineBufferSize, byteCount);
-                            Buffer.BlockCopy(this.batchBuffer.Bytes, this.batchBuffer.CurrentReadPosition, bytesForString, lineBufferSize, byteCount);
-                            lineBufferSize += byteCount;
-                        }
-
-                        this.batchBuffer.SkipTo(lineEndEndPosition + 1);
-                        break;
-                    default:
-                        throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReaderStream_ReadLine));
-                }
-            }
-
-            Debug.Assert(bytesForString != null, "bytesForString != null");
-
-            return this.CurrentEncoding.GetString(bytesForString, 0, lineBufferSize);
-        }
+        internal abstract int ReadWithLength(byte[] userBuffer, int userBufferOffset, int count);
 
         /// <summary>
         /// Ensure that a batch encoding exists; if not, detect it from the first couple of bytes of the stream.
         /// </summary>
-        private void EnsureBatchEncoding()
+        /// <param name="stream">The stream to read.</param>
+        protected void EnsureBatchEncoding(Stream stream)
         {
             // If no batch encoding is specified we detect it from the first bytes in the buffer.
             if (this.batchEncoding == null)
             {
                 // NOTE: The batch encoding will only ever be null on the first call to this method which
                 //       happens before the batch reader skips to the first boundary.
-                this.batchEncoding = this.DetectEncoding();
+                this.batchEncoding = this.DetectEncoding(stream);
             }
 
             // Verify that we only allow single byte encodings and UTF-8 for now.
             ReaderValidationUtils.ValidateEncodingSupportedInBatch(this.batchEncoding);
         }
 
         /// <summary>Detect the encoding based data from the stream.</summary>
+        /// <param name="stream">The stream to read.</param>
         /// <returns>The encoding discovered from the bytes in the buffer or the fallback encoding.</returns>
         /// <remarks>
         /// We don't have to skip a potential preamble of the encoding since the batch reader
         /// will skip over everything (incl. the potential preamble) until it finds the first
         /// boundary.
         /// </remarks>
-        private Encoding DetectEncoding()
+        protected Encoding DetectEncoding(Stream stream)

```

Comment:
>DetectEncoding [](start = 27, length = 14)

Can be moved to multipart/mixed

Reply:
moved.

---
In reply to: [155380791](https://github.com/OData/odata.net/pull/992#discussion_r155380791) [](ancestors = 155380791)